src/pages/resultsView/survival/SurvivalUtil.tsx

```diff
@@ -152,4 +184,87 @@ export function convertScatterDataToDownloadData(patientData: any[]): any[] {
     })
 
     return downloadData;
+}
+
+/**
+ * Down sampling based on the hypotenuse difference.
+ * The cross dot is guaranteed to be drawn when hidden dots are taking into account.
+ * @param {DownSampling} opts
+ * @returns {ScatterData[]}
+ */
+export function downSampling(data: ScatterData[], opts: DownSamplingOpts): ScatterData[] {
+    let x = data.map(item => item.x);
+    let y = data.map(item => item.y);
+
+    let xMax = _.max(x) || 0;
+    let xMin = _.min(x) || 0;
+
+    let yMax = _.max(y) || 0;
+    let yMin = _.min(y) || 0;
+
+    if (opts.xDenominator <= 0 || opts.yDenominator <= 0)
+        return data;
+
+    let timeThreshold = (xMax - xMin) / opts.xDenominator;
+    let survivalRateThreshold = (yMax - yMin) / opts.yDenominator;
+    let averageThreshold = Math.hypot(timeThreshold, survivalRateThreshold);
+    let lastVisibleDataPoint = {
+        x: 0,
+        y: 0
+    };
+    let lastDataPoint = {
+        x: 0,
+        y: 0
+    };
+
+    return data.filter(function (dataItem, index) {
+        let isVisibleDot = _.isUndefined(dataItem.opacity) || dataItem.opacity > 0;
+        if (index == 0) {
+            lastDataPoint.x = dataItem.x;
+            lastDataPoint.y = dataItem.y;
+            lastVisibleDataPoint.x = dataItem.x;
+            lastVisibleDataPoint.y = dataItem.y;
+            return true;
+        }
+        let distance = 0;
+        if (isVisibleDot) {
+            distance = Math.hypot(dataItem.x - lastVisibleDataPoint.x, dataItem.y - lastVisibleDataPoint.y)
+        } else {
+            distance = Math.hypot(dataItem.x - lastDataPoint.x, dataItem.y - lastDataPoint.y);
+        }
+        if (distance > averageThreshold) {
+            lastDataPoint.x = dataItem.x;
+            lastDataPoint.y = dataItem.y;
+            if (isVisibleDot) {
+                lastVisibleDataPoint.x = dataItem.x;
+                lastVisibleDataPoint.y = dataItem.y;
+            }
+            return true;
+        } else {
+            return false;
+        }
+    });
+}
+
+export function filteringScatterData(allScatterData: GroupedScatterData, filters: SurvivalPlotFilters | undefined, downSamplingOpts: DownSamplingOpts):GroupedScatterData {
+    let filteredData = _.cloneDeep(allScatterData);
+    _.forEach(filteredData, (value:SurvivalCurveData, key:string) => {
+        if (value.numOfCases > downSamplingOpts.threshold) {
+            if (filters) {
+                let filteredIndex: number[] = [];
+                _.forEach(value.scatter, (_val, _key) => {
```

Comment:
i think you could do this with a call to reduce that would avoid double iterations (forEach and then filter). and would also be more semantic because you just get assignement  (e.g. result = arr.reduce() )

Reply:
I was trying to reduce the unnecessary loops since this index will be used to get data out of both `.scatter` and `.scatterWithOpacity`. But seems I did it wrong, my intention is using indices to pick up the data instead of `.filter`. But this may not be necessary... after second thought, if most of the dots are qualified the filter, the performance would be worse compared to only two loops. I will change it back.